momepy/functional/_distribution.py

```diff
@@ -152,3 +161,112 @@ def neighbor_distance(geometry: GeoDataFrame | GeoSeries, graph: Graph) -> Serie
+
+    Parameters
+    ----------
+    geometry : GeoDataFrame | GeoSeries
+        A GeoDataFrame or GeoSeries containing geometries to analyse.
+    adjacency_graph : libpysal.graph.Graph
+        Graph representing the adjacency of geometries. Typically, this is a contiguity
+        graph derived from tessellation cells linked to buildings.
+    neighborhood_graph : libpysal.graph.Graph
+        Graph representing the extent around each geometry within which to calculate
+        the mean interbuilding distance. This can be a distance based graph, KNN graph,
+        higher order contiguity, etc.
+
+    Returns
+    -------
+    Series
+    """
+    distance = pd.Series(
+        shapely.distance(
+            geometry.geometry.loc[
+                adjacency_graph._adjacency.index.get_level_values(0)
+            ].values,
+            geometry.geometry.loc[
+                adjacency_graph._adjacency.index.get_level_values(1)
+            ].values,
+        ),
+        index=adjacency_graph._adjacency.index,
+        name="distance",
+    )
+
+    distance_matrix = (
+        distance.astype("Sparse[float]").sparse.to_coo(sort_labels=True)[0].tocsr()
+    )
+    neighborhood_matrix = sparse.coo_matrix(neighborhood_graph.sparse).tocsr()
+
+    mean_distances = np.zeros(distance_matrix.shape[0], dtype=float)
+
+    for i in range(distance_matrix.shape[0]):
+        neighborhood_indices = np.append(neighborhood_matrix[i].indices, i)
+
+        if len(neighborhood_indices) == 0:
+            mean_distances[i] = np.nan
+
+        sub_matrix = distance_matrix[neighborhood_indices][:, neighborhood_indices]
+        mean_distances[i] = sub_matrix.sum() / sub_matrix.nnz
+
+    return Series(
+        mean_distances, index=geometry.index, name="mean_interbuilding_distance"
+    )
+
+
+def building_adjacency(
+    contiguity_graph: Graph,
+    neighborhood_graph: Graph,
+) -> Series:
+    """Calculate the level of building adjacency.
+
+    Building adjacency reflects how much buildings tend to join together into larger
+    structures. It is calculated as a ratio of joined built-up structures captured by
+    ``contiguity_graph`` and buildings within the neighborhood defined in
+    ``neighborhood_graph``.
+
+    Adapted from :cite:`vanderhaegen2017`.
+
+    Notes
+    -----
+    Both graphs must be built on the same index.
+
+    Parameters
+    ----------
+    contiguity_graph : libpysal.graph.Graph
+        Graph representing contiguity between geometries, typically a rook contiguity
+        graph derived from buildings.
+    neighborhood_graph : libpysal.graph.Graph
+        Graph representing the extent around each geometry within which to calculate
+        the level of building adjacency. This can be a distance based graph, KNN graph,
+        higher order contiguity, etc.
+
+    Returns
+    -------
+    Series
+    """
+    components = contiguity_graph.component_labels
+    neighborhood_graph = neighborhood_graph.assign_self_weight()
```

Comment:
[`libpysal==4.8`](https://github.com/pysal/momepy/actions/runs/8209874874/job/22456245168?pr=556#step:3:649) did not have [`Graph().assign_self_weight()`](https://github.com/pysal/libpysal/blob/v4.8.0/libpysal/graph/base.py), which is causing [the failure here](https://github.com/pysal/momepy/actions/runs/8209874874/job/22456245168?pr=556#step:5:429).

Reply:
Were we thinking a skip on `mean_interbuilding_distance()` for now or build in logic for use with `libpysal<4.10`?

Reply:
I would just bump min requirement.